### ISSUE TITLE
Concourse: Using Vault to issue temporary AWS creds

### ISSUE THREAD

**Issue by ghost** [ministryofjustice/cloud-platform#1889]
## Background

This is a follow up to the firebreak ticket of using Vault to issue / revoke dynamic AWS access keys during runtime of those pipelines that provision AWS resources. 

Currently the AWS credentials (both access key / secret ID) used by Concourse  are stored as a Kubernetes secret. 

Below are the main security vulnerabilities with the above approach: 

* Rotating keys is more difficult when they are stored as 'kubernetes secret'

* If the Kubernetes API is comprimised, the AWS keys will be exposed as they can then easily be decoded. (kubernetes secrets are not encrypted, they are simply encoded).

* Having the AWS access keys in pods poses some risks. For example a developer may accidentally log all environment variables. 

## Proposed user journey

Have the ability to avoid having to store AWS access keys as a 'kubernetes secret' resource. Ideally have a scenario where the AWS access keys are created and used on demand for them to be discarded after use. 

## Approach

Add an additional stage to the existing concourse pipeline that interacts with Vault for it to issue a new set of AWS access keys. Concourse will then use these new keys for provisioning the resources. The last step of the pipeline would then revoke the access keys. Most important aspect of this aproach is that the access keys are only ever active during the lifetime of the pipeline. There is no medium that is storing the keys indefinently and persistently. 

Another flexibility to this approach is that different pipelines may use different vault roles that are mapped to AWS policies that are more scoped. The use of which AWS vault role to be used can be parameterised within the pipeline. 

## Which part of the user docs does this impact

This change is low-level enough for it to not affect the current documentation on the high level usage of concourse 

## Questions / Assumptions

Firebreak ticket has already been completed on the POC of using vault to issue temporary AWS access keys, which had been demonstrated to the wider team. 

This ticket is subject to the wider team agreeing to adopting this solution

Below are at least a couple of considerations to this approach: 

* Vault will have to be maintained and managed. Any downtime of vault will block the concourse pipelines from running. However to minismise the risk of downtime and increase recovery time, Vault itself will be running within Kubernetes

* The introduction of a new set of steps in the pipeline that will interact with vault will increase the duration of the pipeline by a few seconds (however long it takes to generate the access keys). 

## Definition of done

- [ ] readme has been updated
- [ ] user docs have been updated
- [ ] another team member has reviewed
- [ ] smoke tests are green

## Reference

[How to write good user stories](https://www.gov.uk/service-manual/agile-delivery/writing-user-stories)

**Comment by digitalronin** [comment]
This is a great ticket, @imranawan-moj Thanks for raising it, and for doing the investigation into Vault.

For me, a very important issue that you raise is this one:

> Vault will have to be maintained and managed. Any downtime of vault will block the concourse pipelines from running.

The cloud platform is already very complex, with lots of moving parts, so I think there would have to be a very compelling benefit before we should take on the additional work of managing any new bit of infrastructure. 

The vulnerabilities you've listed are:

> Rotating keys is more difficult when they are stored as 'kubernetes secret'

I think this is slightly off-target. It's not so much the keys being stored as kubernetes secrets which makes them hard to rotate. IMO the reason our keys are not rotated often is that its tedious to go and generate new ones, not because of the way we store the keys after generating them.

Vault can certainly help with automatic key rotation, but only for certain types of underlying resource, such as AWS and database credentials. It can't do anything about keys it can't generate, e.g. third-party API tokens.

This means we would still have to manage keys which are hard to rotate, although it's true that we could store all of the keys in Vault.

> If the Kubernetes API is comprimised, the AWS keys will be exposed as they can then easily be decoded. (kubernetes secrets are not encrypted, they are simply encoded).

This is true. Someone mentioned during the demo that it is very common for developers to need keys they can use to access resources from their command-line, when troubleshooting or figuring out a problem. If we create a way for developers to do this, then there may still be a way for credentials to be leaked via unauthorised access to the kubernetes API, although it's true that for many resources, those credentials will be so short-lived that this is not much of a problem.

> Having the AWS access keys in pods poses some risks. For example a developer may accidentally log all environment variables.

This is certainly true, and we have had incidents where credentials have been exposed to internal users who should not have access to them.

IMO, there are real benefits in switching to Vault, but right now I would say they're outweighed by the ongoing cost of learning, installing and maintaining yet another piece of critical infrastructure.

An additional 'cost' to think about is changing the way developers work, wrt. secrets. We have quite a lot of experience and know-how spread around the organisation, based on the current way we handle secrets. Changing that requires time and effort by a lot of people, including the time and effort to rewrite our user documentation.

If there were a way to achieve similar results using AWS managed services like KMS+Parameter Store, that could reduce the burden of maintaining an additional component, but it still means more skills to acquire and maintain in the team.

In summary, I think Vault is a great product, and well worth keeping an eye on. If we were building the cloud platform from scratch today, I'd probably use it, but I think the cost of adding it at this stage outweighs the benefits.

As always, that's just my opinion, and I'm very interested to hear the rest of the team's thoughts.

**Comment by mogaal** [comment]
Specifically for our situation, the main advantage that Vault gives us is ephemeral AWS keys for Concourse. A trade-off must be kept in mind: if Vault fails all our pipelines also fail, so it becomes a bottleneck. IMHO for what we get (ephemeral AWS keys) is not worth the maintenance of an extra extra-service. 

Vault's power will be seen when we go multicluster (very related to traffic splitting) because we would be able to maintain secrets in one place (Vault), instead of having multiples copies of `Kind: Secrets`  across different clusters.

I'm not worried about K8S having their secrets in B64 (they're still [encrypted within etcd volume](https://github.com/ministryofjustice/cloud-platform-infrastructure/blob/master/terraform/cloud-platform/templates/kops.yaml.tpl#L184)). I think for what we do this security level is enough if we don't want to sacrifice the maintenance of extra services. 

> Having the AWS access keys in pods poses some risks. For example, a developer may accidentally log all environment variables.

Maybe we are trying to solve the wrong problem here, I think there is no value in having AWS keys at all. We should seriously start considering using roles instead of moving AWS keys from one place to another one. But that is another topic to discuss. 

I've used a lot Vault clusters across my previous positions and I honestly love it. But for what we are trying to solve here I don't see any value. I am sure very soon when we go to traffic splitting, it will be a requirement to have Vault and not an option.

**Comment by imranawan-moj** [comment]
Very good feedback @mogaal / @digitalronin, thanks. Appreciate the cases in not using Vault based on the current setup and perhaps indeed maintaining it would outweigh the benefit. However I must stress the demo I showed makes use of vault roles, which Ale I think you are referring to. The vault role references a AWS policy, and new IAM user is created at the back of this. Keys are only held within the pipeline container during the runtime of the pipeline. So effectively we are not stroing any keys. David - with regards to rotation for keys this actually remninds that it may not even be applicable in this case as we would not even be storing any keys to begin with, as a new fresh set of keys are issued in runtime and then discarded at the end of the pipeline.

**Comment by AntonyBishop** [comment]
Closed due to age of story
